internal/api/container_logs.go

```diff
@@ -7,85 +7,182 @@ package api
+	if err != nil {
+		return fmt.Errorf("getting podName from payload: %w", err)
+	}
 
-			done <- true
-		}()
+	containerName, err := payload.String("containerName")
+	if err != nil {
+		return fmt.Errorf("getting containerName from payload: %w", err)
+	}
 
-		err = container.Logs(r.Context(), kubeClient, namespace, podName, containerName, lines)
-		if err != nil {
-			RespondWithError(w, http.StatusInternalServerError, err.Error(), logger)
-			return
+	eventType := octant.NewLoggingEventType(namespace, podName)
+	val, ok := s.podLogSubscriptions.Load(eventType)
+	if ok {
+		cancelFn, ok := val.(context.CancelFunc)
+		if !ok {
+			return fmt.Errorf("bad cancelFn conversion for %s", eventType)
 		}
+		cancelFn()
+	}
 
-		<-done
+	cancelFn, err := s.createStream(namespace, podName, containerName)
+	if err != nil {
+		return fmt.Errorf("creating log stream: %w", err)
+	}
+
+	s.podLogSubscriptions.Store(eventType, cancelFn)
 
-		var lr logResponse
+	return nil
+}
+
+func (s *podLogsStateManager) StreamPodLogsUnsubscribe(state octant.State, payload action.Payload) error {
+	namespace, err := payload.String("namespace")
+	if err != nil {
+		return fmt.Errorf("getting namespace from payload: %w", err)
+	}
 
-		if len(entries) <= 100 {
-			lr.Entries = entries
-		} else {
-			// take last 100 entries from the slice
-			lr.Entries = entries[len(entries)-100:]
+	podName, err := payload.String("podName")
+	if err != nil {
+		return fmt.Errorf("getting podName from payload: %w", err)
+	}
+
+	eventType := octant.NewLoggingEventType(namespace, podName)
+	val, ok := s.podLogSubscriptions.Load(eventType)
+	if ok {
+		cancelFn, ok := val.(context.CancelFunc)
+		if !ok {
+			return fmt.Errorf("bad cancelFn conversion for %s", eventType)
 		}
+		cancelFn()
+		s.podLogSubscriptions.Delete(eventType)
+	}
+	return nil
+}
+
+func (s *podLogsStateManager) Start(ctx context.Context, state octant.State, client OctantClient) {
+	s.client = client
+	s.ctx = ctx
+}
+
+func (s *podLogsStateManager) createStream(namespace, podName, containerName string) (context.CancelFunc, error) {
+	kubeClient, err := s.config.ClusterClient().KubernetesClient()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancelFn := context.WithCancel(s.ctx)
+	logStreamer, err := container.NewLogStreamer(ctx, kubeClient, namespace, podName, containerName)
+	if err != nil {
+		return cancelFn, err
+	}
 
-		if err := json.NewEncoder(w).Encode(&lr); err != nil {
-			logger := log.From(ctx)
-			logger.With("err", err.Error()).Errorf("unable to encode log entries")
+	eventType := octant.NewLoggingEventType(namespace, podName)
+	logCh := make(chan container.LogEntry)
+
+	go func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case entry, ok := <-logCh:
+				if ok {
+					le := logEntry{
+						Container: entry.Container(),
+						Message:   entry.Line(),
+					}
+					if ts, err := formatTimestamp(le.Message); err != nil {
+						le.Timestamp = &ts
```

Comment:
This should be the opposite. If err is nil, set the timestamp

Reply:
Nice catch!

Reply:
Updated this and added a test.